Tests/GRPCTests/GRPCReflectionServiceTests/GRPCReflectionServiceTests.swift

```diff
@@ -0,0 +1,362 @@
+
+final class GRPCReflectionServiceTests: GRPCTestCase {
+  private var group: MultiThreadedEventLoopGroup?
+  private var server: Server?
+  private var channel: GRPCChannel?
+
+  private func generateProto(name: String, id: Int) -> Google_Protobuf_FileDescriptorProto {
+    let inputMessage = Google_Protobuf_DescriptorProto.with {
+      $0.name = "inputMessage"
+      $0.field = [
+        Google_Protobuf_FieldDescriptorProto.with {
+          $0.name = "inputField"
+          $0.type = .bool
+        }
+      ]
+    }
+
+    let outputMessage = Google_Protobuf_DescriptorProto.with {
+      $0.name = "outputMessage"
+      $0.field = [
+        Google_Protobuf_FieldDescriptorProto.with {
+          $0.name = "outputField"
+          $0.type = .int32
+        }
+      ]
+    }
+
+    let method = Google_Protobuf_MethodDescriptorProto.with {
+      $0.name = "testMethod" + String(id)
+      $0.inputType = inputMessage.name
+      $0.outputType = outputMessage.name
+    }
+
+    let serviceDescriptor = Google_Protobuf_ServiceDescriptorProto.with {
+      $0.method = [method]
+      $0.name = "service" + String(id)
+    }
+
+    let fileDescriptorProto = Google_Protobuf_FileDescriptorProto.with {
+      $0.service = [serviceDescriptor]
+      $0.name = name + String(id) + ".proto"
+      $0.messageType = [inputMessage, outputMessage]
+    }
+
+    return fileDescriptorProto
+  }
+
+  /// Creates the dependencies of the proto used in the testing context.
+  private func makeProtosWithDependencies() -> [Google_Protobuf_FileDescriptorProto] {
+    var fileDependencies: [Google_Protobuf_FileDescriptorProto] = []
+    for id in 1 ... 4 {
+      let fileDescriptorProto = generateProto(name: "bar", id: id)
+      if id != 1 {
+        // Dependency of the first dependency.
+        fileDependencies[0].dependency.append(fileDescriptorProto.name)
+      }
+      fileDependencies.append(fileDescriptorProto)
+    }
+    return fileDependencies
+  }
+
+  private func makeProtosWithComplexDependencies() -> [Google_Protobuf_FileDescriptorProto] {
+    var protos: [Google_Protobuf_FileDescriptorProto] = []
+    protos.append(generateProto(name: "foo", id: 0))
+    for id in 1 ... 10 {
+      var root = protos.count
+      let fileDescriptorProtoA = generateProto(name: "fooA", id: id)
+      let fileDescriptorProtoB = generateProto(name: "fooB", id: id)
+      root = root > 1 ? root - Int.random(in: 1 ..< 3) : root - 1
+      protos[id - 1].dependency.append(fileDescriptorProtoA.name)
+      protos[id - 1].dependency.append(fileDescriptorProtoB.name)
+      protos.append(fileDescriptorProtoA)
+      protos.append(fileDescriptorProtoB)
+    }
+    return protos
+  }
+
+  private func getServicesNamesFromProtos(protos: [Google_Protobuf_FileDescriptorProto]) -> [String]
+  {
+    var servicesNames: [String] = []
+    for proto in protos {
+      servicesNames.append(contentsOf: proto.service.map { $0.name })
+    }
+    return servicesNames
```

Comment:
You can use a `flatMap` here:

```swift
return protos.flatMap { fileDescriptor in 
  fileDescriptor.service.map { $0.name }
}
```

or just `protos.flatMap { $0.service.map { $0.name } }` for short

---

This would actually be nicer as an extension on `Sequence` where the element is `Google_Protobuf_FileDescriptorProto`:

```swift
extension Sequence where Element == Google_Protobuf_FileDescriptorProto {
  var serviceNames: [String] {
    self.flatMap { $0.service.map { $0.name } }
  }
}
```

So then in your tests you can just call `protos.serviceNames`